DOC: Improve pandas.Series.plot.kde docstring and kwargs rewording for whole file

dukebody opened the pull request:
Initial docstring assigned to the Barcelona Chapter for the documentation sprint.

Credits to @arnau126 who also contributed to this change.

Also, rewording of `kwargs` argument explanation for the whole file, where we refer to the `pandas.Series.plot` docstring for info.

dukebody committed:
DOC: Improve pandas.Series.plot.kde docstring.

jorisvandenbossche commented on pandas/plotting/_core.py:
```diff
@@ -2604,26 +2604,73 @@ def hist(self, bins=10, **kwds):
+
+        See also
+        --------
+        :class:`scipy.stats.gaussian_kde` : Representation of a kernel-density
```
Can you check if this works without the ```:class:`..` ```? Normally that should work I think (in the see also section this linking is done automatically)

dukebody replied:
Yes, it works!

jorisvandenbossche commented on pandas/plotting/_core.py:
```diff
@@ -2604,26 +2604,73 @@ def hist(self, bins=10, **kwds):
+            kde is evaluated at the points passed. If `ind` is an integer,
+            `ind` number of equally spaced points are used.
+        kwds : optional
             Keyword arguments to pass on to :py:meth:`pandas.Series.plot`.
```
I am wondering if we can make this more clear, because in the end, what the user needs to know is basically that the other keywords are documented there. 
Would you find something like "Additional keyword arguments are documented in pandas.Series.plot" clearer ?

dukebody replied:
I don't have a strong opinion about the wording, but I think it should be consistent across all docstrings. Right now other docstrings use the `Keyword arguments to pass on to :py:meth:pandas.Series.plot.` wording, so I'd vote for not changing this one now, and decide if we want to change all of them in a row later after the sprint.

What do you think?

jorisvandenbossche replied:
*If* we think it is a better explanation, I would change it now, and then when updating the other docstrings during the sprint, they can take this as an example

dukebody replied:
Hmm I believe this can create a bit of confusion if we have to communicate to everyone in the sprint that this is the docstring they have to take as example and always change the existing wording for the new one, but you probably know better than me. :)

With regards to the explanation itself, know that I've re-read your proposal I believe that yes, it's clearer than the existing one, because the user doesn't need to know that `pandas.Series.plot.kde` calls `pandas.Series.plot` under the hood, just that `pandas.Series.plot.kde` accepts certain keyword arguments and what they mean.

jorisvandenbossche replied:
you can also change it directly everywhere in this PR, that would indeed cause less confusion

dukebody replied:
I can do this if we decide the wording is better. Whose input should we wait for?

I guess it's better to do it in a separate commit and then change the title of the PR. Or perhaps another PR?

jorisvandenbossche replied:
It's fine to do it in this PR. 
And if you and I agree on it, I think it is fine to do the change now (although it can of course be a bit annoying to go back and forth on a change if then other people disagree).

dukebody replied:
Done! I thought this would appear in more files, but it turns out that it's only in this one.

jorisvandenbossche commented on pandas/plotting/_core.py:
```diff
@@ -2604,26 +2604,73 @@ def hist(self, bins=10, **kwds):
+
+        Examples
+        --------
+        Given a ``Series`` of points randomly sampled from an unknown
```
generally in docstrings we don't quote Series or DataFrame (but we don't really have a rule about it)

jorisvandenbossche commented on pandas/plotting/_core.py:
```diff
@@ -2604,26 +2604,73 @@ def hist(self, bins=10, **kwds):
+            :context: close-figs
+
+            >>> s = pd.Series([1, 2, 2.5, 3, 3.5, 4, 5])
+            >>> p = s.plot.kde()
```
The `p = ` is to ensure there is no output (repr of the matplotlib object)? (so the doctests pass?) 
An alternative is to do `s.plot.kde();` (otherwise, I would replace `p` with `ax`, as this is the more common used name for such an object)

dukebody replied:
Yes, exactly! I will use the `ax` alternative.

jorisvandenbossche commented on pandas/plotting/_core.py:
```diff
@@ -2604,26 +2604,73 @@ def hist(self, bins=10, **kwds):
+        .. plot::
+            :context: close-figs
+
+            >>> p = s.plot.kde(bw_method=0.3)
```
maybe switch the order (first small then large value), so it matches the order it is mentioned in the text

jorisvandenbossche reviewed (commented):
Great! Added some comments

dukebody commented:
@jorisvandenbossche Thanks for the review! I've updated the PR in response to your comments.

jorisvandenbossche commented:
There is also issue https://github.com/pandas-dev/pandas/issues/11042 on adding the explanation of more keyword arguments.

dukebody commented:
What about the travis failures? Is it just that the tests are flaky, or the docstring changes can really break something? I.e. are docstring doctests executed by CI?

jorisvandenbossche commented:
> What about the travis failures? Is it just that the tests are flaky, or the docstring changes can really break something? I.e. are docstring doctests executed by CI?

Don't worry about those, we have some failing tests due to the new matplotlib release we need to fix.

dukebody committed:
DOC: Rewording additional kwargs for plot

In the end, what the user needs to know is that the other keywords
are documented in `pandas.Series.plot`. The fact that these plot help
functions call `pandas.Series.plot` under the hood is an
implementation detail.

jorisvandenbossche commented on pandas/plotting/_core.py:
```diff
@@ -2532,7 +2532,8 @@ def line(self, **kwds):
         `**kwds` : optional
-            Keyword arguments to pass on to :py:meth:`pandas.Series.plot`.
+            Additional keyword arguments are documented in
+            :py:meth:`pandas.Series.plot`.
```
sorry for not saying earlier, but you can leave out the `:py` part (that makes it a bit less noisy in the plain text version)

(in principle you can also leave out the 'pandas.' part)

dukebody replied:
Removing the `:py` part but leaving the `pandas.` one.

Related question: What is the difference between `:meth:` and `:class:`? Do they behave differently in docstrings?

jorisvandenbossche replied:
The end goal is to create a link to the help page of that method. But I am not sure how essential it is to use the correct prefix. From experience I know both :func: and :meth: work for methods. But anyway, for the plain text version it is good to use the correct one anyhow (the difference between :meth: and :class: might be how they are displayed in html, with or without brackets)

jorisvandenbossche commented on pandas/plotting/_core.py:
```diff
@@ -2711,7 +2766,8 @@ def line(self, x=None, y=None, **kwds):
         `**kwds` : optional
-            Keyword arguments to pass on to :py:meth:`pandas.DataFrame.plot`.
+            Additional keyword arguments are documented in
+ :py:meth:`pandas.DataFrame.plot`.
```
here went something wrong with the indentation (and some below as well)

dukebody replied:
Sorry, fixing...

dukebody commented:
I believe this should be ready to go. @jorisvandenbossche please do a last review!

shivam6294 commented on pandas/plotting/_core.py:
```diff
@@ -2613,26 +2618,74 @@ def hist(self, bins=10, **kwds):
+            >>> ax = s.plot.kde()
+
+
+        An scalar fixed bandwidth can be specified. Using a too small bandwidth
```
Great work @dukebody! :+1: This is Shivam from the HK Chapter of the Pandas Doc Sprint. 

Just had a minor suggestion for simplifying this paragraph: 
```
        A scalar bandwidth can be specified. Using a small bandwidth value can
        lead to overfitting, while using a large bandwidth value can result 
        in underfitting:
```

jonas-schulze replied:
I agree, @shivam6294 version sounds a little better. Anyways, you have to change `An` to `A` as `scalar` doesn't start with a vowel.

jreback commented on pandas/plotting/_core.py:
```diff
@@ -2613,26 +2618,74 @@ def hist(self, bins=10, **kwds):
+        Kernel Density Estimate plot using Gaussian kernels.
+
+        In statistics, kernel density estimation (KDE) is a non-parametric way
+        to estimate the probability density function (PDF) of a random
```
ideally include a reference to a wiki page (same ones that matplotlib uses?)

jreback commented on pandas/plotting/_core.py:
```diff
@@ -2613,26 +2618,74 @@ def hist(self, bins=10, **kwds):
-            spaced points are used.
-        `**kwds` : optional
-            Keyword arguments to pass on to :py:meth:`pandas.Series.plot`.
+            Evaluation points for the estimated PDF. If None (default),
```
can you spell out PDF

jonas-schulze replied:
Both KDE and DPF are spelled out in the small summary above the parameters section. I wouldn't like to repeat that over here. @dukebody what do you think?

jreback replied:
maybe if you put KDE and PDF in parents then its more clear (do this above)

jreback replied:
nvm I see it.

jreback commented on pandas/plotting/_core.py:
```diff
@@ -2613,26 +2618,74 @@ def hist(self, bins=10, **kwds):
+        --------
+        scipy.stats.gaussian_kde : Representation of a kernel-density
+            estimate using Gaussian kernels. This is the function used
+            internally to estimate the PDF.
```
spell out PDF (can also include PDF in parens)

jonas-schulze commented on pandas/plotting/_core.py:
```diff
@@ -2613,26 +2618,74 @@ def hist(self, bins=10, **kwds):
     def kde(self, bw_method=None, ind=None, **kwds):
         """
-        Kernel Density Estimate plot
+        Kernel Density Estimate plot using Gaussian kernels.
```
The short summary should start with an infinitive verb. I would change it to
```
Draw Kernel Density Estimate plot using Gaussian kernels.
```

jorisvandenbossche replied:
I think for plot functions it is fine to just state the plot type (the rules are to give direction, but sometimes there can be reason to deviate, so in this case the question is if "Draw .." makes it more informative)

jonas-schulze replied:
I agree in that adding `Draw` doesn't add much information, but I would still add a prefix since most (all?) of the doc strings created in the Doc Sprint will be written with these rules in mind. Also, the doc string proposed in [PR20113](https://github.com/pandas-dev/pandas/pull/20113/) for the `hist` function includes `Draw` as well.

On the other hand, `Generate` might be a better fit than `Draw` because out of a jupyter notebook there is nothing drawn ..

jonas-schulze commented on pandas/plotting/_core.py:
```diff
@@ -2613,26 +2618,74 @@ def hist(self, bins=10, **kwds):
         -------
         axes : matplotlib.AxesSubplot or np.array of them
+
+        See also
```
It would be nice to have an additional reference:
```
:meth:`pandas.DataFrame.plot.kde` : Generate a KDE plot for a DataFrame
```

jreback replied:
See also -> See Also

jonas-schulze reviewed (changes requested):
Good work!

jorisvandenbossche commented:
@dukebody Thanks! We will do the follow-up in https://github.com/pandas-dev/pandas/pull/20236
